In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

print("Libraries imported successfully.")

Libraries imported successfully.


In [2]:
#Load the raw dataset
import os

raw_path = "../data/raw"

print("Files available:")
print(os.listdir(raw_path))

Files available:
['delhivery_data.csv']


In [3]:
#Load the CSV
file_path = "../data/raw/delhivery_data.csv"

df = pd.read_csv(file_path)

print("Dataset loaded successfully.")

Dataset loaded successfully.


In [7]:
#Check the dataset size
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])
print(df.columns.tolist())

Rows: 144867
Columns: 24
['data', 'trip_creation_time', 'route_schedule_uuid', 'route_type', 'trip_uuid', 'source_center', 'source_name', 'destination_center', 'destination_name', 'od_start_time', 'od_end_time', 'start_scan_to_end_scan', 'is_cutoff', 'cutoff_factor', 'cutoff_timestamp', 'actual_distance_to_destination', 'actual_time', 'osrm_time', 'osrm_distance', 'factor', 'segment_actual_time', 'segment_osrm_time', 'segment_osrm_distance', 'segment_factor']


In [5]:
#view data
df.head()

,data,trip_creation_time,route_schedule_uuid,route_type,trip_uuid,source_center,source_name,destination_center,destination_name,od_start_time,od_end_time,start_scan_to_end_scan,is_cutoff,cutoff_factor,cutoff_timestamp,actual_distance_to_destination,actual_time,osrm_time,osrm_distance,factor,segment_actual_time,segment_osrm_time,segment_osrm_distance,segment_factor
0,training,2018-09-20 02:35:36.476840,thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...,Carting,trip-153741093647649320,IND388121AAA,Anand_VUNagar_DC (Gujarat),IND388620AAB,Khambhat_MotvdDPP_D (Gujarat),2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,86.0,True,9,2018-09-20 04:27:55,10.435660,14.0,11.0,11.9653,1.272727,14.0,11.0,11.9653,1.272727
1,training,2018-09-20 02:35:36.476840,thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...,Carting,trip-153741093647649320,IND388121AAA,Anand_VUNagar_DC (Gujarat),IND388620AAB,Khambhat_MotvdDPP_D (Gujarat),2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,86.0,True,18,2018-09-20 04:17:55,18.936842,24.0,20.0,21.7243,1.200000,10.0,9.0,9.7590,1.111111
2,training,2018-09-20 02:35:36.476840,thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...,Carting,trip-153741093647649320,IND388121AAA,Anand_VUNagar_DC (Gujarat),IND388620AAB,Khambhat_MotvdDPP_D (Gujarat),2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,86.0,True,27,2018-09-20 04:01:19.505586,27.637279,40.0,28.0,32.5395,1.428571,16.0,7.0,10.8152,2.285714
3,training,2018-09-20 02:35:36.476840,thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...,Carting,trip-153741093647649320,IND388121AAA,Anand_VUNagar_DC (Gujarat),IND388620AAB,Khambhat_MotvdDPP_D (Gujarat),2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,86.0,True,36,2018-09-20 03:39:57,36.118028,62.0,40.0,45.5620,1.550000,21.0,12.0,13.0224,1.750000
4,training,2018-09-20 02:35:36.476840,thanos::sroute:eb7bfc78-b351-4c0e-a951-fa3d5c3...,Carting,trip-153741093647649320,IND388121AAA,Anand_VUNagar_DC (Gujarat),IND388620AAB,Khambhat_MotvdDPP_D (Gujarat),2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,86.0,False,39,2018-09-20 03:33:55,39.386040,68.0,44.0,54.2181,1.545455,6.0,5.0,3.9153,1.200000


In [6]:
df.tail()

,data,trip_creation_time,route_schedule_uuid,route_type,trip_uuid,source_center,source_name,destination_center,destination_name,od_start_time,od_end_time,start_scan_to_end_scan,is_cutoff,cutoff_factor,cutoff_timestamp,actual_distance_to_destination,actual_time,osrm_time,osrm_distance,factor,segment_actual_time,segment_osrm_time,segment_osrm_distance,segment_factor
144862,training,2018-09-20 16:24:28.436231,thanos::sroute:f0569d2f-4e20-4c31-8542-67b86d5...,Carting,trip-153746066843555182,IND131028AAB,Sonipat_Kundli_H (Haryana),IND000000ACB,Gurgaon_Bilaspur_HB (Haryana),2018-09-20 16:24:28.436231,2018-09-20 23:32:09.618069,427.0,True,45,2018-09-20 21:57:20,45.258278,94.0,60.0,67.9280,1.566667,12.0,12.0,8.1858,1.000000
144863,training,2018-09-20 16:24:28.436231,thanos::sroute:f0569d2f-4e20-4c31-8542-67b86d5...,Carting,trip-153746066843555182,IND131028AAB,Sonipat_Kundli_H (Haryana),IND000000ACB,Gurgaon_Bilaspur_HB (Haryana),2018-09-20 16:24:28.436231,2018-09-20 23:32:09.618069,427.0,True,54,2018-09-20 21:31:18,54.092531,120.0,76.0,85.6829,1.578947,26.0,21.0,17.3725,1.238095
144864,training,2018-09-20 16:24:28.436231,thanos::sroute:f0569d2f-4e20-4c31-8542-67b86d5...,Carting,trip-153746066843555182,IND131028AAB,Sonipat_Kundli_H (Haryana),IND000000ACB,Gurgaon_Bilaspur_HB (Haryana),2018-09-20 16:24:28.436231,2018-09-20 23:32:09.618069,427.0,True,63,2018-09-20 21:11:18,66.163591,140.0,88.0,97.0933,1.590909,20.0,34.0,20.7053,0.588235
144865,training,2018-09-20 16:24:28.436231,thanos::sroute:f0569d2f-4e20-4c31-8542-67b86d5...,Carting,trip-153746066843555182,IND131028AAB,Sonipat_Kundli_H (Haryana),IND000000ACB,Gurgaon_Bilaspur_HB (Haryana),2018-09-20 16:24:28.436231,2018-09-20 23:32:09.618069,427.0,True,72,2018-09-20 20:53:19,73.680667,158.0,98.0,111.2709,1.612245,17.0,27.0,18.8885,0.629630
144866,training,2018-09-20 16:24:28.436231,thanos::sroute:f0569d2f-4e20-4c31-8542-67b86d5...,Carting,trip-153746066843555182,IND131028AAB,Sonipat_Kundli_H (Haryana),IND000000ACB,Gurgaon_Bilaspur_HB (Haryana),2018-09-20 16:24:28.436231,2018-09-20 23:32:09.618069,427.0,False,70,2018-09-20 16:24:28.436231,70.039010,426.0,95.0,88.7319,4.484211,268.0,9.0,8.8088,29.777778


Check data types

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144867 entries, 0 to 144866
Data columns (total 24 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   data                            144867 non-null  object 
 1   trip_creation_time              144867 non-null  object 
 2   route_schedule_uuid             144867 non-null  object 
 3   route_type                      144867 non-null  object 
 4   trip_uuid                       144867 non-null  object 
 5   source_center                   144867 non-null  object 
 6   source_name                     144574 non-null  object 
 7   destination_center              144867 non-null  object 
 8   destination_name                144606 non-null  object 
 9   od_start_time                   144867 non-null  object 
 10  od_end_time                     144867 non-null  object 
 11  start_scan_to_end_scan          144867 non-null  float64
 12  is_cutoff       

Check missing values

In [9]:
missing = df.isnull().sum()

missing[missing > 0].sort_values(ascending=False)

source_name         293
destination_name    261
dtype: int64

Check duplicate rows

In [10]:
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


Basic statistics

In [11]:
df.describe(include="all").T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
data,144867,2,training,104858,NaN,NaN,NaN,NaN,NaN,NaN,NaN
trip_creation_time,144867,14817,2018-09-28 05:23:15.359220,101,NaN,NaN,NaN,NaN,NaN,NaN,NaN
route_schedule_uuid,144867,1504,thanos::sroute:4029a8a2-6c74-4b7e-a6d8-f9e069f...,1812,NaN,NaN,NaN,NaN,NaN,NaN,NaN
route_type,144867,2,FTL,99660,NaN,NaN,NaN,NaN,NaN,NaN,NaN
trip_uuid,144867,14817,trip-153811219535896559,101,NaN,NaN,NaN,NaN,NaN,NaN,NaN
source_center,144867,1508,IND000000ACB,23347,NaN,NaN,NaN,NaN,NaN,NaN,NaN
source_name,144574,1498,Gurgaon_Bilaspur_HB (Haryana),23347,NaN,NaN,NaN,NaN,NaN,NaN,NaN
destination_center,144867,1481,IND000000ACB,15192,NaN,NaN,NaN,NaN,NaN,NaN,NaN
destination_name,144606,1468,Gurgaon_Bilaspur_HB (Haryana),15192,NaN,NaN,NaN,NaN,NaN,NaN,NaN
od_start_time,144867,26369,2018-09-21 18:37:09.322207,81,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Let's inspect the missing records

In [12]:
df[df["source_name"].isna()][
    ["source_center", "source_name", "destination_center", "destination_name"]
].head(20)

,source_center,source_name,destination_center,destination_name
112,IND342902A1B,NaN,IND302014AAA,Jaipur_Hub (Rajasthan)
113,IND342902A1B,NaN,IND302014AAA,Jaipur_Hub (Rajasthan)
114,IND342902A1B,NaN,IND302014AAA,Jaipur_Hub (Rajasthan)
115,IND342902A1B,NaN,IND302014AAA,Jaipur_Hub (Rajasthan)
116,IND342902A1B,NaN,IND302014AAA,Jaipur_Hub (Rajasthan)
117,IND342902A1B,NaN,IND302014AAA,Jaipur_Hub (Rajasthan)
118,IND342902A1B,NaN,IND302014AAA,Jaipur_Hub (Rajasthan)
119,IND342902A1B,NaN,IND302014AAA,Jaipur_Hub (Rajasthan)
120,IND342902A1B,NaN,IND302014AAA,Jaipur_Hub (Rajasthan)
121,IND342902A1B,NaN,IND302014AAA,Jaipur_Hub (Rajasthan)


In [13]:
df[df["destination_name"].isna()][
    ["source_center", "source_name", "destination_center", "destination_name"]
].head(20)

,source_center,source_name,destination_center,destination_name
110,IND342601AAA,Piparcity_BsstdDPP_D (Rajasthan),IND342902A1B,NaN
111,IND342601AAA,Piparcity_BsstdDPP_D (Rajasthan),IND342902A1B,NaN
982,IND573103AAA,Arsikere_HsnRdDPP_D (Karnataka),IND577116AAA,NaN
983,IND573103AAA,Arsikere_HsnRdDPP_D (Karnataka),IND577116AAA,NaN
4882,IND202001AAB,Aligarh_KhirByps_I (Uttar Pradesh),IND282002AAD,NaN
4883,IND202001AAB,Aligarh_KhirByps_I (Uttar Pradesh),IND282002AAD,NaN
4884,IND202001AAB,Aligarh_KhirByps_I (Uttar Pradesh),IND282002AAD,NaN
4885,IND202001AAB,Aligarh_KhirByps_I (Uttar Pradesh),IND282002AAD,NaN
7246,IND465001AAA,Shajapur_MhpraRD_D (Madhya Pradesh),IND465333A1B,NaN
7247,IND465001AAA,Shajapur_MhpraRD_D (Madhya Pradesh),IND465333A1B,NaN


Then check whether the missing names are associated with unique centers

In [14]:
print("Unique source centers with missing source_name:")
print(df.loc[df["source_name"].isna(), "source_center"].nunique())

print("\nUnique destination centers with missing destination_name:")
print(df.loc[df["destination_name"].isna(), "destination_center"].nunique())

Unique source centers with missing source_name:
10

Unique destination centers with missing destination_name:
13


Data Cleaning

In [46]:
# Check all records for centers with missing source names

missing_source_centers = df.loc[
    df["source_name"].isna(),
    "source_center"
].unique()

source_mapping_check = (
    df[df["source_center"].isin(missing_source_centers)]
    .groupby("source_center")["source_name"]
    .unique()
)

source_mapping_check

source_center
IND126116AAA    [nan]
IND282002AAD    [nan]
IND331022A1B    [nan]
IND342902A1B    [nan]
IND465333A1B    [nan]
IND505326AAB    [nan]
IND509103AAC    [nan]
IND577116AAA    [nan]
IND841301AAC    [nan]
IND852118A1B    [nan]
Name: source_name, dtype: object

In [16]:
# Check all records for centers with missing destination names

missing_destination_centers = df.loc[
    df["destination_name"].isna(),
    "destination_center"
].unique()

destination_mapping_check = (
    df[df["destination_center"].isin(missing_destination_centers)]
    .groupby("destination_center")["destination_name"]
    .unique()
)

destination_mapping_check

destination_center
IND122015AAC    [nan]
IND126116AAA    [nan]
IND221005A1A    [nan]
IND250002AAC    [nan]
IND282002AAD    [nan]
IND331001A1C    [nan]
IND342902A1B    [nan]
IND465333A1B    [nan]
IND505326AAB    [nan]
IND509103AAC    [nan]
IND577116AAA    [nan]
IND841301AAC    [nan]
IND852118A1B    [nan]
Name: destination_name, dtype: object

Handle missing center names

In [18]:
# Create a copy of the raw DataFrame first:
df_clean = df.copy()

In [19]:
df_clean["source_name"] = df_clean["source_name"].fillna("Unknown")
df_clean["destination_name"] = df_clean["destination_name"].fillna("Unknown")

In [20]:
print("Missing source_name:",
      df_clean["source_name"].isna().sum())

print("Missing destination_name:",
      df_clean["destination_name"].isna().sum())

Missing source_name: 0
Missing destination_name: 0


Check ALL remaining missing values

In [21]:
missing_after_names = df_clean.isnull().sum()

missing_after_names[missing_after_names > 0].sort_values(ascending=False)

Series([], dtype: int64)

Check data types

In [23]:
df_clean.dtypes

data                               object
trip_creation_time                 object
route_schedule_uuid                object
route_type                         object
trip_uuid                          object
source_center                      object
source_name                        object
destination_center                 object
destination_name                   object
od_start_time                      object
od_end_time                        object
start_scan_to_end_scan            float64
is_cutoff                            bool
cutoff_factor                       int64
cutoff_timestamp                   object
actual_distance_to_destination    float64
actual_time                       float64
osrm_time                         float64
osrm_distance                     float64
factor                            float64
segment_actual_time               float64
segment_osrm_time                 float64
segment_osrm_distance             float64
segment_factor                    

In [24]:
# Then specifically check the columns that should represent dates/times:
date_columns = [
    "trip_creation_time",
    "od_start_time",
    "od_end_time",
    "cutoff_timestamp"
]

df_clean[date_columns].head()

,trip_creation_time,od_start_time,od_end_time,cutoff_timestamp
0,2018-09-20 02:35:36.476840,2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,2018-09-20 04:27:55
1,2018-09-20 02:35:36.476840,2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,2018-09-20 04:17:55
2,2018-09-20 02:35:36.476840,2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,2018-09-20 04:01:19.505586
3,2018-09-20 02:35:36.476840,2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,2018-09-20 03:39:57
4,2018-09-20 02:35:36.476840,2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,2018-09-20 03:33:55


In [25]:
df_clean[date_columns].dtypes

trip_creation_time    object
od_start_time         object
od_end_time           object
cutoff_timestamp      object
dtype: object

#. Data Cleaning

Data cleaning is performed to ensure the dataset is accurate, consistent, and ready for exploratory analysis and downstream modeling.

The main cleaning tasks include:

- Handling missing values
- Checking duplicate records
- Converting date/time columns to the appropriate data type
- Validating data consistency
- Preparing the dataset for feature engineering and analysis

## Converting Date/Time Columns

The dataset contains four columns representing timestamps:

- `trip_creation_time`
- `od_start_time`
- `od_end_time`
- `cutoff_timestamp`

These columns are initially stored as `object` (text) data types.

For time-based analysis, they need to be converted to Pandas `datetime` format.

In [26]:
date_columns = [
    "trip_creation_time",
    "od_start_time",
    "od_end_time",
    "cutoff_timestamp"
]

for col in date_columns:
    df_clean[col] = pd.to_datetime(df_clean[col], errors="coerce")

### Verify Datetime Conversion

We verify the data types after conversion to ensure that all timestamp columns are correctly recognized as datetime values.

In [27]:
df_clean[date_columns].dtypes

trip_creation_time    datetime64[ns]
od_start_time         datetime64[ns]
od_end_time           datetime64[ns]
cutoff_timestamp      datetime64[ns]
dtype: object

### Check for Invalid or Unconverted Timestamps

The conversion uses `errors="coerce"`, which converts any invalid timestamp values into `NaT` (Not a Time).

We therefore check for newly created missing datetime values after conversion.

In [28]:
df_clean[date_columns].isna().sum()

trip_creation_time       0
od_start_time            0
od_end_time              0
cutoff_timestamp      3429
dtype: int64

In [ ]:
df_clean.dtypes

data                               object
trip_creation_time                 object
route_schedule_uuid                object
route_type                         object
trip_uuid                          object
source_center                      object
source_name                        object
destination_center                 object
destination_name                   object
od_start_time                      object
od_end_time                        object
start_scan_to_end_scan            float64
is_cutoff                            bool
cutoff_factor                       int64
cutoff_timestamp                   object
actual_distance_to_destination    float64
actual_time                       float64
osrm_time                         float64
osrm_distance                     float64
factor                            float64
segment_actual_time               float64
segment_osrm_time                 float64
segment_osrm_distance             float64
segment_factor                    

### Observation

All timestamp columns were successfully converted to datetime format without introducing additional missing values.

In [29]:
pd.to_datetime(..., errors="coerce")

NaT

##  Investigating Invalid Cutoff Timestamps

After converting the timestamp columns to datetime format, the `cutoff_timestamp` column contains 3,429 `NaT` values.

These values may have been caused by inconsistent or invalid timestamp formats in the raw dataset.

Before deciding how to handle them, we inspect the corresponding raw values and compare them with other available time fields.

In [30]:
cutoff_invalid = df_clean[df_clean["cutoff_timestamp"].isna()]

print("Invalid cutoff timestamps:", len(cutoff_invalid))

cutoff_invalid[
    [
        "trip_creation_time",
        "od_start_time",
        "od_end_time",
        "cutoff_timestamp",
        "is_cutoff",
        "cutoff_factor"
    ]
].head(20)

Invalid cutoff timestamps: 3429


,trip_creation_time,od_start_time,od_end_time,cutoff_timestamp,is_cutoff,cutoff_factor
2,2018-09-20 02:35:36.476840,2018-09-20 03:21:32.418600,2018-09-20 04:47:45.236797,NaT,True,27
53,2018-09-12 01:33:48.711350,2018-09-12 06:12:09.579013,2018-09-12 07:19:24.177399,NaT,True,36
54,2018-09-12 01:33:48.711350,2018-09-12 06:12:09.579013,2018-09-12 07:19:24.177399,NaT,False,37
60,2018-09-12 01:33:48.711350,2018-09-12 07:19:24.177399,2018-09-12 09:26:45.211239,NaT,False,56
66,2018-09-12 01:33:48.711350,2018-09-12 09:26:45.211239,2018-09-12 11:26:24.365997,NaT,False,52
86,2018-09-13 16:41:41.992300,2018-09-13 18:54:06.314808,2018-09-13 20:40:01.943322,NaT,False,79
106,2018-09-25 08:53:04.377810,2018-09-25 21:31:00.741991,2018-09-26 00:21:56.684666,NaT,False,61
111,2018-09-25 08:53:04.377810,2018-09-26 05:04:49.254901,2018-09-26 06:58:08.054001,NaT,False,34
203,2018-09-17 00:14:20.789064,2018-09-17 01:59:09.679228,2018-09-17 03:19:31.471358,NaT,False,37
280,2018-10-03 23:24:57.864330,2018-10-03 23:24:57.864330,2018-10-04 00:21:58.203266,NaT,True,22


In [31]:
cutoff_invalid["is_cutoff"].value_counts(dropna=False)

is_cutoff
False    2841
True      588
Name: count, dtype: int64

In [32]:
cutoff_invalid["cutoff_factor"].describe()

count    3429.000000
mean       54.815981
std       112.579412
min         9.000000
25%        22.000000
50%        33.000000
75%        52.000000
max      1721.000000
Name: cutoff_factor, dtype: float64

## Handling Missing Cutoff Timestamps

After converting `cutoff_timestamp` to datetime format, 3,429 records were converted to `NaT`.

The affected records include both:

- 2,841 records where `is_cutoff = False`
- 588 records where `is_cutoff = True`

Because the original timestamp information cannot be reliably reconstructed from the available dataset, these values are retained as `NaT` rather than being replaced with an arbitrary timestamp.

The associated `is_cutoff` and `cutoff_factor` fields are preserved because they still contain useful operational information.

This approach avoids introducing fabricated data while retaining the affected logistics records for further analysis.

In [33]:
cutoff_summary = pd.DataFrame({
    "Metric": [
        "Total records",
        "Missing cutoff timestamps",
        "Missing cutoff timestamps (%)"
    ],
    "Value": [
        len(df_clean),
        df_clean["cutoff_timestamp"].isna().sum(),
        round(df_clean["cutoff_timestamp"].isna().mean() * 100, 2)
    ]
})

cutoff_summary

,Metric,Value
0,Total records,144867.00
1,Missing cutoff timestamps,3429.00
2,Missing cutoff timestamps (%),2.37


## Numeric Column Validation

The dataset contains several numerical variables related to delivery performance, distance, and route segments.

These columns will be validated to ensure they are stored using appropriate numeric data types before performing statistical analysis and feature engineering.

In [34]:
numeric_columns = [
    "start_scan_to_end_scan",
    "cutoff_factor",
    "actual_distance_to_destination",
    "actual_time",
    "osrm_time",
    "osrm_distance",
    "factor",
    "segment_actual_time",
    "segment_osrm_time",
    "segment_osrm_distance",
    "segment_factor"
]

df_clean[numeric_columns].dtypes

start_scan_to_end_scan            float64
cutoff_factor                       int64
actual_distance_to_destination    float64
actual_time                       float64
osrm_time                         float64
osrm_distance                     float64
factor                            float64
segment_actual_time               float64
segment_osrm_time                 float64
segment_osrm_distance             float64
segment_factor                    float64
dtype: object

### Numeric Data Validation

The following check confirms whether the selected logistics metrics are stored as numeric variables and are therefore suitable for statistical calculations and feature engineering.

In [38]:
df_clean[numeric_columns].select_dtypes(include="number").columns.tolist()

['start_scan_to_end_scan',
 'cutoff_factor',
 'actual_distance_to_destination',
 'actual_time',
 'osrm_time',
 'osrm_distance',
 'factor',
 'segment_actual_time',
 'segment_osrm_time',
 'segment_osrm_distance',
 'segment_factor']

##  Categorical Column Validation

Several columns describe categorical characteristics of the delivery operation, such as route type, cutoff status, and data partition.

These variables will be reviewed for their data types and unique values before exploratory analysis.

In [40]:
categorical_columns = [
    "data",
    "route_type",
    "trip_uuid",
    "source_center",
    "source_name",
    "destination_center",
    "destination_name",
    "is_cutoff"
]

df_clean[categorical_columns].dtypes

data                  object
route_type            object
trip_uuid             object
source_center         object
source_name           object
destination_center    object
destination_name      object
is_cutoff               bool
dtype: object

### Unique Values in Key Categorical Variables

The number of unique values is checked to understand the structure and cardinality of the categorical variables.

In [41]:
for col in ["data", "route_type", "is_cutoff"]:
    print(f"\n{col}:")
    print(df_clean[col].value_counts(dropna=False))


data:
data
training    104858
test         40009
Name: count, dtype: int64

route_type:
route_type
FTL        99660
Carting    45207
Name: count, dtype: int64

is_cutoff:
is_cutoff
True     118749
False     26118
Name: count, dtype: int64


In [42]:
print("Unique trip_uuid:", df_clean["trip_uuid"].nunique())
print("Unique source centers:", df_clean["source_center"].nunique())
print("Unique destination centers:", df_clean["destination_center"].nunique())

Unique trip_uuid: 14817
Unique source centers: 1508
Unique destination centers: 1481


### Dataset Granularity

The dataset contains 144,867 records but only 14,817 unique trips identified by `trip_uuid`.

This indicates that multiple records can belong to the same trip. Therefore, the dataset contains information at a more granular level than the overall trip.

The dataset also contains 1,508 unique source centers and 1,481 unique destination centers.

For analysis, the appropriate level of aggregation will depend on the business question. Segment-level analysis will be used for detailed route performance, while `trip_uuid` will be used to aggregate records when evaluating overall trip performance.

##  Categorical Data Summary

The dataset contains two route types: `FTL` and `Carting`, with FTL representing the majority of records.

The `data` column separates the dataset into training and test records. This original designation is preserved for downstream analysis and modeling.

The `is_cutoff` variable is dominated by `True` records. This imbalance will be considered when analyzing cutoff behavior and when developing predictive models.

In [44]:
print("Dataset shape:", df_clean.shape)
print("Duplicate rows:", df_clean.duplicated().sum())
print("\nMissing values:")
print(df_clean.isnull().sum()[df_clean.isnull().sum() > 0])

Dataset shape: (144867, 24)
Duplicate rows: 0

Missing values:
cutoff_timestamp    3429
dtype: int64


In [45]:
processed_path = "../data/processed/delhivery_cleaned.csv"

df_clean.to_csv(processed_path, index=False)

print(f"Cleaned dataset saved to: {processed_path}")

Cleaned dataset saved to: ../data/processed/delhivery_cleaned.csv


## Saving the Cleaned Dataset

The cleaned dataset is saved separately from the raw dataset.

The original raw data remains unchanged in `data/raw/`, while the cleaned version is stored in `data/processed/`.

This preserves data lineage and allows the original dataset to be reproduced or reprocessed if required.

Exploratory Data Analysis